File: epi_judge_python/is_tree_a_bst.py
from test_framework import generic_test


def is_binary_tree_bst(tree, low_range=float('-inf'), high_range=float('inf')):
    if tree is None:
        return True
    if is_binary_tree_bst(tree.left, low_range, high_range= tree.data) \
        and is_binary_tree_bst(tree.right, low_range=tree.data, high_range=high_range) \
        and low_range <= tree.data <= high_range:
        return True
    return False


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("is_tree_a_bst.py", 'is_tree_a_bst.tsv',
                                       is_binary_tree_bst))

File: epi_judge_java/epi/IsTreeABst.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class IsTreeABst {
  private static class BSTStatus{
    boolean isBST;
    int min;
    int max;

    public BSTStatus(boolean isBST, int min, int max){
      this.isBST = isBST;
      this.min = min;
      this.max = max;
    }
  }
  public static BSTStatus statusChecker(BinaryTreeNode<Integer> u){
    int min = u.data, max = u.data;
    boolean isBST = true;
    if (u.left != null){
      BSTStatus leftStatus = statusChecker(u.left);
      isBST = isBST && leftStatus.isBST && leftStatus.max <= u.data;
      min = Math.min(min, leftStatus.min);
    }
    if (u.right != null){
      BSTStatus rightStatus = statusChecker(u.right);
      isBST = isBST && rightStatus.isBST && rightStatus.min >= u.data;
      max = Math.max(max, rightStatus.max);
    }

    return new BSTStatus(isBST, min,max);
  }
  @EpiTest(testDataFile = "is_tree_a_bst.tsv")

  public static boolean isBinaryTreeBST(BinaryTreeNode<Integer> tree) {
    // TODO - you fill in here.
    if (tree == null){
      return true;
    }
    return statusChecker(tree).isBST;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsTreeABst.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
